mava/systems/sable/sebulba/rec_sable.py

```diff
@@ -0,0 +1,787 @@
+
+                obs_tpu = tree.map(move_to_device, timestep.observation)
+
+                prev_dones = jnp.repeat(timestep.last(), config.system.num_agents).reshape(
+                    config.arch.num_envs, -1
+                )
+
+                # Get action and value
+                with RecordTimeTo(actor_timings["compute_action_time"]):
+                    key, act_key = jax.random.split(key)
+                    action, log_prob, value, hstates = act_fn(params, obs_tpu, hstates, act_key)
+                    cpu_action = jax.device_get(action)
+
+                # Step environment
+                with RecordTimeTo(actor_timings["env_step_time"]):
+                    timestep = env.step(cpu_action)
+
+                # Updated the dones and Hstates
+                dones = timestep.last()
+                dones = jnp.expand_dims(dones, (1, 2, 3, 4))
+                hstates = tree.map(
+                    lambda hs, dones=dones: jnp.where(dones, jnp.zeros_like(hs), hs), hstates
+                )
+
+                # Append data to storage
+                traj.append(
+                    Transition(prev_dones, action, value, timestep.reward, log_prob, obs_tpu)
+                )
+                episode_metrics.append(timestep.extras["episode_metrics"])
+
+        # send trajectories to learner
+        with RecordTimeTo(actor_timings["rollout_put_time"]):
+            try:
+                rollout_queue.put(
+                    traj, (actor_timings, episode_metrics), (timestep, (prev_hstates, hstates))
+                )
+            except queue.Full:
+                err = "Waited too long to add to the rollout queue, killing the actor thread"
+                warnings.warn(err, stacklevel=2)
+                break
+
+    env.close()
+
+
+def get_learner_step_fn(
+    apply_fns: Tuple[ActorApply, LearnerApply],
+    update_fn: optax.TransformUpdateFn,
+    config: DictConfig,
+) -> Callable[[LearnerState, Transition, HiddenStates, HiddenStates], Tuple[LearnerState, Metrics]]:
+    """Get the learner function."""
+
+    num_envs = config.arch.num_envs
+    num_learner_envs = int(num_envs // len(config.arch.learner_device_ids))
+
+    # Get apply and update functions for actor and critic networks.
+    sable_action_select_fn, sable_apply_fn = apply_fns
+
+    def _update_step(
+        learner_state: LearnerState,
+        traj_batch: Transition,
+        prev_hstates: HiddenStates,
+        updated_hstates: HiddenStates,
+    ) -> Tuple[LearnerState, Metrics]:
+        """A single update of the network.
+
+        This function calculates advantages and targets based on the trajectories
+        from the actor and updates the actor and critic networks based on the losses.
+
+        Args:
+            learner_state (LearnerState): contains all the items needed for learning.
+            traj_batch (PPOTransition): the batch of data to learn with.
+        """
```

Comment:
This was quite confusing to me at first glance, can we change it to this (please also change where ever we use these names e.g: rollout and minibatching -  might be useful to change it in the anakin system also :pray:)

```suggestion
        initial_hstate: HiddenStates,
        final_hstate: HiddenStates,
    ) -> Tuple[LearnerState, Metrics]:
        """A single update of the network.

        This function calculates advantages and targets based on the trajectories
        from the actor and updates the actor and critic networks based on the losses.

        Args:
            learner_state (LearnerState): contains all the items needed for learning.
            traj_batch (PPOTransition): the batch of data to learn with.
            initial_hstate (HiddenState): the hidden state from the start of the rollout.
            final_hstate (HiddenState): the last hidden state of the rollout.
        """
```